### PR TITLE
Dev loll

### PR DESCRIPTION
**Pull request by lollcat**
First version of adding a replay buffer to improve sample efficiency, partially solves #6 .

- The current replay buffer saves log weights and samples generated by AIS - which are then re-used for multiple updates to the flow. The efficiency of this could be further improved by saving the final value of the target log probability within AIS, and re-using this also - however I have left this for a further update. 
- Currently this just works with the `alpha_2_loss`, and not other losses 
- A new trainer is created for training with the replay buffer, as the structure of training is a bit different (saving & re-using samples). 

This MR also includes some updates from my branch that are unrelated to the buffer.
-  Form of the forward kl loss (using self-normalised weights), which improved it's stability in practice. 
-  minor miscellaneous updates to notebooks

### CODE REVIEW

File: fab/core.py [hunk 1]
```diff
@@ -74,8 +80,9 @@ def fab_forward_kl(self, batch_size: int) -> torch.Tensor:
         x_ais, log_w_ais = self.annealed_importance_sampler.sample_and_log_weights(batch_size)
         x_ais = x_ais.detach()
         log_w_ais = log_w_ais.detach()
+        w_ais = torch.softmax(log_w_ais, dim=0)
         log_q_x = self.flow.log_prob(x_ais)
-        return - torch.mean(torch.exp(log_w_ais) * log_q_x)
+        return - torch.mean(w_ais * log_q_x)
```
Comment (lollcat) [inline fab/core.py:85]: This is not related to the buffer MR, but I am including it from my dev branch as it improved the stability of the forward kl loss.

File: fab/train_with_buffer.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            info = self.model.get_iter_info()
+            info.update(loss=loss.cpu().detach().item(),
+                        step=i,
+                        grad_norm=grad_norm.cpu().detach().item())
+            self.logger.write(info)
+
+            # We now take an additinal self.n_batches_buffer_sampling gradient steps using
+            # data from the replay buffer.
+            for (x, log_w) in self.buffer.sample_n_batches(
+                    batch_size=batch_size, n_batches=self.n_batches_buffer_sampling):
```
Comment (lollcat) [inline fab/train_with_buffer.py:105]: @VincentStimper these lines (72 - 119) will need to be equivalently updated for the dipeptide problem

Comment (lollcat) [reply fab/train_with_buffer.py:105]: We unfortunately can no longer have a simple loop where we repeatedly just call `loss()`

**Review by VincentStimper — APPROVED** [review]
Looks very good. Thanks a lot for your effort!
